vault/ssorolecredentialsprovider.go

```diff
@@ -201,3 +211,191 @@ func (p *SSORoleCredentialsProvider) newOIDCToken(ctx context.Context) (*ssooidc
+	if err != nil || n != 32 {
+		return nil, fmt.Errorf("failed to generate PKCE verifier: %w", err)
+	}
+	codeVerifier := base64.RawURLEncoding.EncodeToString(codeVerifierBytes)
+
+	// generate the code challenge: base64(sha256(codeVerifier))
+	codeChallengeBytes := sha256.Sum256([]byte(codeVerifier))
+	codeChallenge := base64.RawURLEncoding.EncodeToString(codeChallengeBytes[:])
+	log.Printf("Generated PKCE code_challenge: %q", codeChallenge)
+
+	clientCreds, err := p.OIDCClient.RegisterClient(ctx, &ssooidc.RegisterClientInput{
+		ClientName:   aws.String("aws-vault"),
+		ClientType:   aws.String("public"),
+		GrantTypes:   []string{"authorization_code", "refresh_token"},
+		Scopes:       []string{"sso:account:access"},
+		IssuerUrl:    aws.String(p.StartURL),
+		RedirectUris: []string{"http://127.0.0.1/oauth/callback"},
+	})
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("Created new OIDC client (expires at: %s)", time.Unix(clientCreds.ClientSecretExpiresAt, 0))
+
+	// start the callback server
+	cbServer, err := newOauthCallbackServer()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create oauthCallbackServer: %w", err)
+	}
+	log.Printf("oauthCallbackServer callback endpoint: %s", cbServer.redirectURI())
+	go func() {
+		if err := cbServer.Serve(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Printf("Failed to run oauthCallbackServer: %s", err)
+		}
+	}()
+	// keep a copy of the redirectURI for the CreateToken call (as the server will be closed after the code is received)
+	redirectURI := cbServer.redirectURI()
+
+	// construct the authorize URL with the client and PKCE parameters
+	args := url.Values{
+		"client_id":             {aws.ToString(clientCreds.ClientId)},
+		"response_type":         {"code"},
+		"redirect_uri":          {redirectURI},
+		"state":                 {cbServer.state},
+		"code_challenge_method": {"S256"},
+		"code_challenge":        {codeChallenge},
+		"scopes":                {"sso:account:access"},
+	}
+	// prefer the base endpoint from client options, otherwise use a default
+	var host string
+	if p.OIDCClient.Options().BaseEndpoint != nil && *p.OIDCClient.Options().BaseEndpoint != "" {
+		host = *p.OIDCClient.Options().BaseEndpoint
+	} else {
+		host = "oidc.us-east-1.amazonaws.com"
+	}
+	authorizeURL := url.URL{
+		Scheme:   "https",
+		Host:     host,
+		Path:     "/authorize",
+		RawQuery: args.Encode(),
+	}
+	log.Printf("Authorize URL: %s", authorizeURL.String())
+
+	// redirect user to the authorize URL
+	log.Println("Opening SSO authorization page in browser")
+	fmt.Fprintf(os.Stderr, "Opening the SSO authorization page in your default browser (use Ctrl-C to abort)\n%s\n", authorizeURL.String())
+	if err := open.Run(authorizeURL.String()); err != nil {
+		log.Printf("Failed to open browser: %s", err)
+	}
+
+	// await the authorization code from the callback server once the user has completed the flow.
+	var code string
+	timeout := time.After(1 * time.Minute)
```

Comment:
FWIW the device code implementation [here](https://github.com/99designs/aws-vault/blob/949040e0413c23ac97ecdd1f397680643a294e76/vault/ssorolecredentialsprovider.go#L181) waits indefinitely (modulo a timeout on the AWS side), so perhaps it's worth increasing it?